pybart/new_matcher.py

```diff
@@ -0,0 +1,330 @@
+# usage example:
+# conversions = [SomeConversion, SomeConversion1, ...]
+# matcher = Matcher(NamedConstraint(conversion.get_name(), conversion.get_constraint()) for conversion in conversions)
+#
+# def convert_sentence(sentence, matcher) -> sentence:
+#     while … # till convergance of the sentence
+#         m = matcher(doc)
+#         for conv_name in m.names():
+#             matches = m.matches_for(conv_name)
+#             transform(matches…) # TBD
+#         sentence = ...
+#
+# for doc in docs:
+#     convert_sentence(sentence, matcher)
+
+import itertools
+from collections import defaultdict
+from typing import NamedTuple, Sequence, Mapping, Any, List, Tuple
+import re
+import spacy
+from spacy.matcher import Matcher as SpacyMatcher
+from .constraints import *
+
+
+# function that checks that a sequence of Label constraints is satisfied
+# TODO - maybe move this function as a 'satisfied' method of the Label constraint? will simplify this code to
+#   "for constraint in label_constraints: if not label_constraint.satisfied(actual_labels): return False; return True"
+def are_labels_satisfied(label_constraints: Sequence[Label], actual_labels: List[str]) -> Tuple[bool, Set[str]]:
+    successfully_matched = set()
+    # we need to satisfy all constraints in the sequence, so if one fails, return False
+    for constraint in label_constraints:
+        if isinstance(constraint, HasNoLabel):
+            # check if a regex or exact match is required
+            is_regex = constraint.value.startswith('/') and constraint.value.endswith('/')
```

Comment:
the regex-vs-value decision should be done when creating the constraint, not each time it is invoked.